docs/src/main/sphinx/develop/client-protocol.rst

```diff
@@ -0,0 +1,257 @@
+
+If the client request returns an HTTP 503, that means the server was busy,
+and the client should try again in 50-100 milliseconds.  Any HTTP status other
+than 503 or 200 means that the query has failed.
+
+The ``/v1/statement`` ``POST`` request returns a JSON document of type
+``QueryResults``, as well as a collection of response headers.  The
+``QueryResults`` document will contain an ``error`` field of type
+``QueryError`` if the query has failed, and if that object is not present,
+the query succeeded.  Important members of ``QueryResults`` are documented
+below.
+
+If the ``data`` field of the JSON document is set, it contains a list of the
+rows of data, and the ``columns`` field will also be set to a list of the
+names and types of the columns returned by the query.  Most of the response
+headers should be treated like browser cookies by the client, and echoed back
+as request headers in subsequent client requests, as documented below.
+
+If the JSON document returned by the ``POST`` to ``/v1/statement`` does not
+contain a ``nextUri`` link, the query has completed, either successfully or
+unsuccessfully, and no additional requests need to be made.  If the
+``nextUri`` link is present in the document, there are more query results
+to be fetched.  The client should loop executing a ``GET`` request
+to the ``nextUri`` returned in the ``QueryResults`` response object until
+``nextUri`` is absent from the response.
+
+The ``status`` field of the JSON document is for human consumption only, and
+provides a hint about the query's state on the server.  It is not synchronized
+with the server's query state and should not be used to tell if the query
+is finished.
+
+Important ``QueryResults`` attributes
+=====================================
+
+The most important attributes of the ``QueryResults`` JSON document returned by the REST API
+endpoints are listed in this table.  Refer to the class ``QueryResults`` for more details.
+
+.. list-table:: ``QueryResults attributes``
+  :widths: 25, 55
+  :header-rows: 1
+
+  * - Attribute
+    - Description
+  * - ``id``
+    - The ID of the query.
+  * - ``nextUri``
+    - If present, the URL that should be used in subsequent ``GET`` or
+      ``DELETE`` requests.  If not present, the query is complete or
+      ended in error.
+  * - ``columns``
+    - A list of the names and types of the columns returned by the query.
+  * - ``data``
+    - The ``data`` attribute contains a list of the rows returned by the
+      query request.  Each row is itself a list that holds values of the
+      columns in the row, in the order specified by the ``columns``
+      attribute.
+  * - ``updateType``
+    - A human-readable string representing the operation.  For a
+      ``CREATE TABLE`` request, the ``updateType`` will be
+      "CREATE TABLE"; for ``SET SESSION`` it will be "SET SESSION"; etc.
+  * - ``error``
+    - If query failed, the ``error`` attribute will contain JSON for a
```

Comment:
If the query failed, the error attribute contains a QueryError object

Reply:
I think the fact that it is JSON encoded is obvious/irrelevant

Reply:
Agreed; changed.